src/writeckpt.cpp

```diff
@@ -467,84 +459,69 @@ mtcp_write_non_rwx_and_anonymous_pages(int fd, Area *orig_area)
     area.addr += size;
     area.size -= size;
   }
-
-  /* Now remove the PROT_READ from the area if it didn't have it originally
-  */
-  if ((orig_area->prot & PROT_READ) == 0) {
-    JASSERT(mprotect(orig_area->addr, orig_area->size, orig_area->prot) == 0)
-      (JASSERT_ERRNO) ((void*)orig_area->addr) (orig_area->size)
-    .Text("error removing PROT_READ from mem region.");
-  }
 }
 
 static void
-writememoryarea(int fd, Area *area, int stack_was_seen)
+writememoryarea(int fd, Area area)
 {
   int rc = 0;
-  void *addr = area->addr;
+  void *addr = area.addr;
 
-  if (!(area->flags & MAP_ANONYMOUS)) {
-    JTRACE("save region") (addr) (area->size) (area->name) (area->offset);
-  } else if (area->name[0] == '\0') {
-    JTRACE("save anonymous") (addr) (area->size);
+  if (!(area.flags & MAP_ANONYMOUS)) {
+    JTRACE("save region") (addr) (area.size) (area.name) (area.offset);
+  } else if (area.name[0] == '\0') {
+    JTRACE("save anonymous") (addr) (area.size);
   } else {
-    JTRACE("save anonymous") (addr) (area->size) (area->name) (area->offset);
+    JTRACE("save anonymous") (addr) (area.size) (area.name) (area.offset);
   }
 
-  if ((area->name[0]) == '\0') {
+  if ((area.name[0]) == '\0') {
     char *brk = (char *)sbrk(0);
-    if (brk > area->addr && brk <= area->addr + area->size) {
-      strcpy(area->name, "[heap]");
+    if (brk > area.addr && brk <= area.addr + area.size) {
+      strcpy(area.name, "[heap]");
     }
   }
 
-  if (area->size == 0) {
-    /* Kernel won't let us munmap this.  But we don't need to restore it. */
-    JTRACE("skipping over [stack] segment (not the orig stack)")
-      (addr) (area->size);
-  } else if (area->prot == 0 ||
-             (area->name[0] == '\0' &&
-              ((area->flags & MAP_ANONYMOUS) != 0) &&
-              ((area->flags & MAP_PRIVATE) != 0))) {
+  if (area.name[0] == '\0') {
```

Comment:
Isn't it a limiting condition? These regions can be named `/anon_hugepage`.

Reply:
Good catch. Added MAP_Anonymous check as well. It might still be limiting and we’ll probably need something like !IsRegularFile()

Reply:
Thanks, but yes, it is still limiting.
A simple dmtcp1 ckpt-restart failing (core dumped with SIGBUS) on my machine with the following backtrace:
```
#0  0x00007f8ef4e098a8 in dmtcp::Util::areZeroPages (addr=0x7f8ef393b000, numPages=255) at util_misc.cpp:577
#1  0x00007f8ef4de6e7d in mtcp_get_next_page_range (area=0x7f8ef15eadc0, size=0x7f8ef15ebdc8, is_zero=0x7f8ef15ebdc4) at writeckpt.cpp:414
#2  0x00007f8ef4de7168 in mtcp_write_anonymous_pages (fd=5, area=...) at writeckpt.cpp:441
#3  0x00007f8ef4de756b in writememoryarea (fd=5, area=...) at writeckpt.cpp:490
#4  0x00007f8ef4de6655 in mtcp_writememoryareas (fd=5) at writeckpt.cpp:352
#5  0x00007f8ef4dc8077 in dmtcp::CkptSerializer::writeCkptImage (mtcpHdr=0x7f8ef15f2120, mtcpHdrLen=4096, 
    ckptFilename="/home/twinkle/dmtcp_3_0/dmtcp/ckpt_dmtcp1_5c0352704509e818-40000-42401362c5931.dmtcp.temp") at ckptserializer.cpp:450
#6  0x00007f8ef4dded8c in dmtcp::ThreadList::writeCkpt () at threadlist.cpp:310
#7  0x00007f8ef4ddf067 in checkpointhread (dummy=0x0) at threadlist.cpp:410
#8  0x00007f8ef4de2a75 in thread_start (arg=0x7f8ef5de0008) at threadwrappers.cpp:70
#9  0x00007f8ef43754f9 in start_thread (arg=0x7f8ef15f4700) at pthread_create.c:465
#10 0x00007f8ef488bf2f in clone () at ../sysdeps/unix/sysv/linux/x86_64/clone.S:95
```
```
(gdb) whereis-address 0x7f8ef383b000
/lib64/librt-2.26.so 0x7f8ef383b000-0x7f8ef3a3a000 (0x7f8ef3a3a000)
```
It's failing because DMTCP is trying to checkpoint the library's protected region. The earlier check for protection would avoid checking zero pages/segments in protected regions.